### ISSUE TITLE
Port C++ marked source rendering code to Java. [phab:D1980]

### ISSUE THREAD

**Issue by kythe** [google/kythe#2433]
_Created by **zarko** at 2017-10-17 17:05:13_:

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2017-10-17 17:45:47_:

In all of these, I don't think the (new OutsideParens(stuff)) is necessary, you can just do new WithoutParens(stuff).render(...

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2017-10-17 17:45:47_:

real nit here, but any sense in making the ordering of params consistent (makelink, signature) or (signature, makelink), pick one and use in both places?

**Comment by kythe** [comment]
_**danielmoy-google** commented inline (context lost during migration) at 2017-10-17 17:45:47_:

That is worth a doc for the poor saps reading this in N years.

I'm not opposed to bit magic, but who knows who will be looking at this later.

Actually while we're here, I wonder if it makes more sense to have constants for your bit positions, so you don't need to remember which one is 5 down here.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-10-17 19:22:52_:

I would prefer to replace all of the magic with an `EnumSet` instead.

**Comment by kythe** [comment]
_**schroederc** required changes at 2017-10-17 19:22:52_

**Comment by kythe** [comment]
_**zarko** wrote at 2017-10-17 21:16:36_:

  - Address comments.

**Comment by kythe** [comment]
_**zarko** wrote at 2017-10-17 21:19:50_:

  - Split working branch

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-10-17 22:20:11_:

I don't think these static sets are warranted.  Their uses would be more readable as the plain `EnumSet.of(...)` calls.

In either case, you should use `Sets.immutableEnumSet` instead.

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-10-17 22:20:11_:

Revert?

**Comment by kythe** [comment]
_**schroederc** commented inline (context lost during migration) at 2017-10-17 22:20:11_:

This would probably be better without the fall-through `switch` now.  Maybe something like this?

```
MarkedSource.Kind kind = node.getKind();
if (!MarkedSource.Kind.BOX.equals(kind)) {
  if (!MarkedSource.Kind.IDENTIFIER.equals(kind) && !enabled.contains(kind)) {
    return;
  }
  under = Stream.concat(under.stream(), Stream.of(kind)).collect(Sets.toImmutableEnumSet());
}
```

**Comment by kythe** [comment]
_**zarko** wrote at 2017-10-17 22:53:12_:

  - Address comments.

**Comment by kythe** [comment]
_**schroederc** accepted this revision at 2017-10-17 23:02:15_

**Comment by kythe** [comment]
_**zarko** closed via commit 97b8204081a290d5e99b471e3935dc56a9cf0824 at 2017-10-17 23:49:02_
